### PR TITLE
Make Xrdb Loading More Forgiving

### PR DESCRIPTION
**Pull request by a-kenji**
Attempt at fixing  #434.

The `hex_to_rgb` function propagates the Option now.
If it is None, it tries to load the default Palette in most cases.

### CODE REVIEW

File: src/common/utils/shared.rs [hunk 1]
```diff
@@ -43,12 +43,33 @@ pub mod colors {
     pub const BLACK: (u8, u8, u8) = (0, 0, 0);
 }
 
-pub fn hex_to_rgb(hex: &Option<String>) -> (u8, u8, u8) {
+pub fn hex_to_rgb(hex: &Option<String>) -> Option<(u8, u8, u8)> {
```
Comment (yykamei) [inline src/common/utils/shared.rs:46]: How about removing `Option`? Because `hex_to_rgb` seems to be a util function, I wonder it doesn't have to care about the argument is `Some` or not.

By the way, `Rgb::from_hex_str` returns `Result`, but this function looks like ignoring it with `unwrap`. I think it's OK because `hex_str` is usually a valid color hex in many cases. However, using `expect` might be more helpful for developers to detect errors.

To summarize, I came up with this:

```rust
pub fn hex_to_rgb(hex: &str) -> (u8, u8, u8) {
    let rgb = Rgb::from_hex_str(hex)
        .expect("The passed argument should be valid color hex.")
        .as_tuple();
    (rgb.0 as u8, rgb.1 as u8, rgb.2 as u8)
}
```

**Review by yykamei — COMMENTED** [review]
LGTM. I like this approach. It makes sense to return with `default_palette()` though some users might use their own terminal with a light background color.

File: src/common/os_input_output.rs [hunk 2]
```diff
@@ -268,10 +268,23 @@ impl OsApi for OsInputOutput {
     fn load_palette(&self) -> Palette {
         let palette = match Colors::new("xresources") {
             Some(palette) => {
-                let fg = hex_to_rgb(&palette.fg);
-                let bg = hex_to_rgb(&palette.bg);
-                let colors: Vec<(u8, u8, u8)> =
-                    palette.colors.iter().map(|c| hex_to_rgb(c)).collect();
+                let fg = if let Some(foreground) = hex_to_rgb(&palette.fg) {
```
Comment (yykamei) [inline src/common/os_input_output.rs:271]: If `hex_to_rgb` should be like [this](https://github.com/zellij-org/zellij/pull/437/files#r624784691) comment, this line might be fixed with the following change (`palette.bg` is also fixed as the same change :smile:)

```suggestion
                let fg = if let Some(foreground) = palette.fg.as_deref().map(hex_to_rgb) {
```

File: src/common/os_input_output.rs [hunk 3]
```diff
@@ -268,10 +268,23 @@ impl OsApi for OsInputOutput {
...
+                let bg = if let Some(background) = hex_to_rgb(&palette.bg) {
+                    background
+                } else {
+                    return default_palette();
+                };
+
+                let colors: Vec<(u8, u8, u8)> = palette
+                    .colors
+                    .iter()
+                    .map(|c| hex_to_rgb(c).unwrap())
```
Comment (yykamei) [inline src/common/os_input_output.rs:286]: :thinking: In my environment, all elements of `palette.colors` are `None`, so it might be better to consider the case `c` is `None`.

File: src/common/utils/shared.rs (hunk 1 above)
Comment (a-kenji) [reply src/common/utils/shared.rs:46]: Thank you, yes that makes a lot of sense. Your solution is very elegant!

File: src/common/os_input_output.rs (hunk 2 above)
Comment (a-kenji) [reply src/common/os_input_output.rs:271]: Yes, that makes sense :+1: .

File: src/common/os_input_output.rs (hunk 3 above)
Comment (yykamei) [reply src/common/os_input_output.rs:286]: It might be ugly but practical to use `for` instead of `iter()`. This will get fallback to the default when any item of `palette.colors` is `None`.

```rust
let mut colors: [(u8, u8, u8); 16] = [(0, 0, 0); 16]; // NOTE: `16` is the same as the length of `palette.colors`.
for (idx, color) in palette.colors.iter().enumerate() {
    if let Some(c) = color {
        colors[idx] = hex_to_rgb(c);
    } else {
        return default_palette();
    }
}
```

**Comment by denismaxim0v** [comment]
LGTM

**Review by denismaxim0v — APPROVED** [review]

File: src/common/os_input_output.rs (hunk 3 above)
Comment (a-kenji) [reply src/common/os_input_output.rs]: Its not as nice as the `iter()`, but much nicer and more concise as I came up with, since we sadly couldn't use a closure here.

**Comment by a-kenji** [comment]
@yykamei 
Thanks for checking it over and leaving so much good advice!
> It makes sense to return with `default_palette()` though some users might use their own terminal with a light background color.

I hope we will soon-ish be able to have theming options in the config.
